### PR TITLE
Slider: Enforce numeric constraints and styling within the text input

### PR DESCRIPTION
**Pull request by drew08t**
What this PR does / why we need it:
Improved state management is needed when using one instance of slider across multiple layers, for example in geomap.

- [x] Add NumberInput to core slider
- [x] Improve state management
- [x] Improve styling for NumberInput adjacent to slider
- [x] Simplify

### CODE REVIEW

File: public/app/core/components/OptionsUI/slider.tsx [hunk 1]
```diff
@@ -1,26 +1,85 @@
...
+          defaultValue={value}
+          value={sliderValue}
+          onChange={onSliderChange}
+          vertical={!isHorizontal}
+          reverse={false}
+          ariaLabelForHandle={ariaLabelForHandle}
+          marks={marks}
+          included={included}
+        />
+        {/* Uses text input so that the number spinners are not shown */}
```
Comment (ryantxu) [inline public/app/core/components/OptionsUI/slider.tsx:80]: except on hover... and despite all my efforts, will take a change in grafana/ui to fix that

File: public/app/core/components/OptionsUI/slider.tsx [hunk 2]
```diff
@@ -1,26 +1,85 @@
...
+        <NumberInput value={sliderValue} onChange={onSliderInputChange} max={settings?.max} min={settings?.min} />
```
Comment (ryantxu) [inline public/app/core/components/OptionsUI/slider.tsx:81]:
```suggestion
        <NumberInput value={sliderValue} onChange={onSliderInputChange} max={max} min={min} step={step} />
```
the number values should match the slider -- if the slider is auto clamped to 0-100, so should the text

**Comment by drew08t** [comment]
Slider + NumberInput styling:
Jun-16-2022 20-05-39

File: public/app/core/components/OptionsUI/slider.tsx [hunk 3]
```diff
@@ -85,8 +106,36 @@ export const SliderValueEditor: React.FC<FieldConfigEditorProps<number, SliderFi
...
+
+const getStylesSlider = (theme: GrafanaTheme2, width: number) => {
+  return {
+    numberInputWrapper: css`
+      margin-left: 10px;
+      max-width: ${width}px;
+      min-width: ${width}px;
+      width: 100%;
+    `,
+    numberInputHidden: css`
```
Comment (drew08t) [inline public/app/core/components/OptionsUI/slider.tsx:137]: Need to remove this.

File: public/app/core/components/OptionsUI/slider.tsx [hunk 4]
```diff
@@ -1,26 +1,141 @@
...
+        <span className={stylesSlider.numberInputWrapper} ref={inputRef}>
+          <NumberInput value={sliderValue} onChange={onSliderInputChange} max={max} min={min} step={step} />
+        </span>
+      </label>
+    </div>
   );
 };
+
+// Calculate width of string with given font
+function getTextWidth(text: string, font: string): number | null {
```
Comment (ryantxu) [inline public/app/core/components/OptionsUI/slider.tsx:118]: is `measureText` not good enough?

Comment (drew08t) [reply public/app/core/components/OptionsUI/slider.tsx:118]: I am using that below on line 123, but it requires context to return the TextMetrics object.

File: public/app/core/components/OptionsUI/slider.tsx [hunk 5]
```diff
@@ -1,26 +1,143 @@
...
+import { Global } from '@emotion/react';
+import SliderComponent from 'rc-slider';
+import React, { useCallback, useEffect, useRef, useState } from 'react';
 
-import { FieldConfigEditorProps, SliderFieldConfigSettings } from '@grafana/data';
-import { Slider } from '@grafana/ui';
+import { FieldConfigEditorProps, GrafanaTheme2, SliderFieldConfigSettings } from '@grafana/data';
+import { useTheme2 } from '@grafana/ui';
+
+import { getStyles } from '../../../../../packages/grafana-ui/src/components/Slider/styles';
```
Comment (ryantxu) [inline public/app/core/components/OptionsUI/slider.tsx:9]:
```suggestion
import { getStyles } from '@grafana/ui/src/components/Slider/styles';
```

**Comment by ryantxu** [comment]
Is this ready for review yet?

**Comment by ryantxu** [comment]
Here is some kinda funky behavior... but not really sure how to best improve it:

**Review by ryantxu — APPROVED** [review]

**Comment by ryantxu** [comment]
Since this has "add-to-changelog" we should make sure the title is meaningful :) maybe:

`Slider: Enforce numeric constraints within the text input`. or something

**Comment by drew08t** [comment]
> Here is some kinda funky behavior... but not really sure how to best improve it:  [ ](https://user-images.githubusercontent.com/705951/177646549-e0a4bbbc-57e3-4e34-af73-66180e35f8fb.gif)  [ ](https://user-images.githubusercontent.com/705951/177646549-e0a4bbbc-57e3-4e34-af73-66180e35f8fb.gif)

The warning that pops up and impacts the nearby element styles?

**Comment by drew08t** [comment]
Looks like the styling behavior could be addressed here in Field.tsx:
image

**Comment by drew08t** [comment]
@ryantxu If you like this styling instead, I can commit it. Best I could do without going deeper and modifying form Field.

Jul-06-2022 18-34-05

**Pull request merged**